### ISSUE TITLE
Acceptance Threshold on a record task

### ISSUE THREAD

**Issue by cmgrier** [cashapp/paparazzi#722]
I am requesting a task similar to `recordPaparazziDebug` that records new and failing snapshots that are above a minimum threshold of variance (similarly to how the verify task catches differences). If the current snapshot passes verification, this task won't rerecord that snapshot.

For context, I work in a project that has lots of snapshots (in the hundreds) and when even one fails, we see that sometimes all snapshots are rerecorded, even when the verify task would only catch the one failing snapshot. This makes it very difficult to see what is actually changed in PRs.

**Comment by Stofferoo** [comment]
I've run into this problem as well now. I guess the best way of making this happen is to have a verify task run and only execute those tests that failed?

**Comment by agrosner** [comment]
we have this issue at our company, where we may have hundreds of snapshots. We have to run `recordPaparazziDebug` at a module level (which is totally fine). However, we have each test output the junit parameterized product of phone / tablet x 5 brand themes = 10 snaps per test case. Some modules have well over 100 snapshots in them. Running record will account for the error drift and rerecord them, showing up as a diff in github. 

What we would love to have, is a tast like `verifyOrRecordPaparazziDebug` which will only re-record snapshots that fail. To account for real changes that have happened in a PR. Otherwise the real changes get lost in a set of identical looking side-by-side images. 🙏

**Comment by RowanG1** [comment]
Is there any update on this request?

**Comment by matejdro** [comment]
We came up with a script to do this thing automatically. It essentially copies all current images to a different folder, runs record task and then copies back all images that did not fail.

It only works on a single module, but I think it could be modified to work on multiple.


```bash
#!/usr/bin/env bash
set -e

SCRIPT_DIR=$(pwd)
# Script assumes it's in the APP_DIR/scripts folder, so it only needs to move one folder up to get to the root. Modify this if needed.
cd ..

echo "Running Paparazzi to get a list of all failed snapshots (following Gradle task SHOULD fail, don't worry)..."
echo ""
(./gradlew -q :app:verifyPaparazzi && (echo "All snapshots already pass, nothing to generate"; exit 0)) || true

# Parse a list of all failed snaphosts (we only need to generate these)
FAILED_SNAPSHOTS=$(cat app/build/test-results/testDebugUnitTest/TEST-com.myapp.PaparazziTest.xml | grep -B 1 "<failure" | grep "testcase" | grep -Po "(?<=test\[).*(?=\])")

# Copy original snapshots to a temp directory
cp -r app/src/test/snapshots/images $SCRIPT_DIR/snapshots

echo ""
echo "Generating new snapshots..."
echo ""
./gradlew -q :app:recordPaparazzi

# Copy unchanged snapshots back
for SNAPSHOT in $SCRIPT_DIR/snapshots/images/*.*; do
  IS_FAILED=false
  for FAILED_SNAPSHOT in $FAILED_SNAPSHOTS; do
    if echo "$SNAPSHOT" | grep -q "$FAILED_SNAPSHOT"; then
      IS_FAILED=true
      break
    fi
  done

  if [[ $IS_FAILED == false ]]; then
    cp $SNAPSHOT app/src/test/snapshots/images/
  else
    echo "Detected updated snapshot: $SNAPSHOT"
  fi
done

# Cleanup
rm -rf $SCRIPT_DIR/snapshots
```

**Comment by minkuan88** [comment]
When running `verifyPaparazzi`, it seems like the failed snapshots are being recorded into `build/paparazzi/failures`. Simply running a find and copy after verify seems to work for me

```sh
./gradlew verifyPaparazzi || exit_code=$?

if [ $exit_code -ne 0 ]; then
  find build/paparazzi/failures -type f -name "*.png" -not -name "delta-*.png" -print -exec cp {} src/test/snapshots/images/ \;
fi
``

**Comment by matejdro** [comment]
Good idea, this one is much shorter. Although, it will only update failing snapshots, not the new ones.
